src/server/frontend_wayland/wlr_screencopy_v1.cpp

```diff
@@ -134,14 +219,73 @@ mf::WlrScreencopyManagerV1::WlrScreencopyManagerV1(
 {
 }
 
+mf::WlrScreencopyManagerV1::~WlrScreencopyManagerV1()
+{
+    if (change_notifier)
+    {
+        ctx->surface_stack->remove_observer(change_notifier);
+    }
+}
+
+void mf::WlrScreencopyManagerV1::maybe_wait_for_damage(FrameParams const& params, WlrScreencopyFrameV1* frame)
+{
+    if (!change_notifier)
+    {
+        // We create the change notifier the first time a client requests a frame with damage
+        create_change_notifier();
+    }
+    auto const area = std::find_if(
+        begin(capture_areas),
+        end(capture_areas),
+        [&](auto area){ return area.params == params; });
+    if (area != end(capture_areas))
+    {
+        area->add_frame(frame);
+    }
+    else
+    {
+        // We capture the given frame immediately, and also push an empty capture area so that if we get another
+        // capture request with the same params it will wait for damage since this frame.
+        frame->capture(std::nullopt);
+        capture_areas.emplace_back(params);
+        // If an unusually high number of capture areas have been created for some reason, clear the list rather than
+        // getting bogged down (it's ok, worst case scenario waiting for damage doesn't work)
+        if (capture_areas.size() > 100)
```

Comment:
Can capture areas ever be cancelled? Is there really a problem with large numbers? Should we simply stop adding them instead of clearing the existing ones?

Reply:
My reading of the protocol is that captures are requested a buffer at a time. If I have that right `capture_areas` is acting both as a list of current captures and as a cache of previous ones? (The cache is obviously sensible as requests for damage only make sense when the same parameters are used for repeated requests)

The right way to invalidate the cache in this scenario would be dropping the least recently used entries.

Reply:
Indeed making it a deque and popping from the front might be optimal, but needing any sort of cache invalidation is unlikely so I just went with the simplest thing that would keep it from getting bogged down.